src/main/java/io/jenkins/plugins/opentelemetry/job/OtelTraceService.java

```diff
@@ -218,27 +219,36 @@ public void purgeRun(@NonNull Run run) {
 
     public void putSpan(@NonNull AbstractBuild build, @NonNull Span span) {
         build.addAction(new MonitoringAction(span));
+        build.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
         LOGGER.log(Level.FINEST, () -> "putSpan(" + build.getFullDisplayName() + "," + OtelUtils.toDebugString(span) + ")");
     }
 
     public void putSpan(AbstractBuild build, BuildStep buildStep, Span span) {
         build.addAction(new BuildStepMonitoringAction(span));
+        build.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
         LOGGER.log(Level.FINEST, () -> "putSpan(" + build.getFullDisplayName() + ", " + buildStep + "," + OtelUtils.toDebugString(span) + ")");
     }
 
     public void putSpan(@NonNull Run run, @NonNull Span span) {
         run.addAction(new MonitoringAction(span));
+        run.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
         LOGGER.log(Level.FINEST, () -> "putSpan(" + run.getFullDisplayName() + "," + OtelUtils.toDebugString(span) + ")");
     }
 
     public void putRunPhaseSpan(@NonNull Run run, @NonNull Span span) {
         run.addAction(new RunPhaseMonitoringAction(span));
+        run.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
         LOGGER.log(Level.FINEST, () -> "putRunPhaseSpan(" + run.getFullDisplayName() + "," + OtelUtils.toDebugString(span) + ")");
     }
 
     public void putSpan(@NonNull Run run, @NonNull Span span, @NonNull FlowNode flowNode) {
         // FYI for agent allocation, we have 2 FlowNodeMonitoringAction to track the agent allocation duration
         flowNode.addAction(new FlowNodeMonitoringAction(span));
+        run.getActions(AttributeSetterAction.class).forEach(setter -> setter.setToSpan(span));
```

Comment:
This could be a problem on large pipelines with hundreds of children. Also, siblings in a deep tree will iterate over and over again over the upper levels.

Reply:
I changed this to use SpanContext instead.
Only for some spans (agent allocation, finalise phase) a single `OpenTelemetryAttributesAction` on the Run is used for getting the attributes.